### PR TITLE
feat: Support mermaid flowchart and sequence diagrams to excalidraw diagrams 🥳

### PR DESCRIPTION
**Pull request by ad1992**
https://github.com/excalidraw/excalidraw/assets/11256141/db111fea-6660-4769-bf4d-b8d0ea47feac

```[tasklist]
### TODOS
- [x] Integrate mermaid-to-excalidraw
- [x] Fix export bug - for labelled arrows the export is breaking
- [x] markdown syntax is breaking with the mermaid package
- [x] export preview should be centered
- [x] Mark elements as selected when inserted into canvas
- [x] Export preview should never go out of the preview box
- [x] Load mermaid-to-excalidraw lazilly
- [x] Fix package preview as webpack doesn't recognize module type
- [x] Design tweaks
- [x] Make it work in mobile
- [x] Fix tests
- [x] Add tests
- [x] Persist data to local storage
- [x] Vertical labelled arrow issue
- [x] Undefined id issue
- [x] error handling
- [x] Error message in React Way
- [x] Fix arrow bindings in Safari and Firefox
- [x] Use scene coords instead of view port coords when pasting to canvas
```

### CODE REVIEW

**Comment by ad1992** [comment]
@excalibot trigger release

**Comment by zsviczian** [comment]
This is very nice. Do you think there is a way to convert the Mermaid diagram into Excalidraw elements?

**Comment by ad1992** [comment]
> This is very nice. Do you think there is a way to convert the Mermaid diagram into Excalidraw elements?

@zsviczian we have written our own parser for it (will make the repo public soon)
for now you can try out in preview (I have added a video in desc for the same)

**Comment by dwelle** [comment]
🎉

Haven't checked code yet, but some notes on the UI:

image

**Comment by dwelle** [comment]
It seems the mermaid chunks are super granular, and also being loaded right on init load, not when the dialog is being opened.

Also we're not splitting our vendor or other chunks too greedily, resulting in too many requests (83 vs 45 in production)

image

**Comment by zsviczian** [comment]
This is super exciting!

It's a small thing, but I would do two things:
- In the description at the top of the dialog window I would add a link to mermaid documentation for those who are unfamiliar mermaid: https://mermaid.js.org/intro/n00b-gettingStarted.html 
- and I would include a simple sample flowchart that people can start to tweak (or delete), instead of greeting them with an empty notepad and an empty preview.

**Comment by ad1992** [comment]
> It seems the mermaid chunks are super granular, and also being loaded right on init load, not when the dialog is being opened.
> 
> Also we're not splitting our vendor or other chunks too greedily, resulting in too many requests (83 vs 45 in production)
> 
> image

As discussed in chat previously, I will be doing it in separate PR (testing with mermaid changes) and merge that to master first.

**Comment by ad1992** [comment]
> This is super exciting!
> 
> It's a small thing, but I would do two things:
> 
> * In the description at the top of the dialog window I would add a link to mermaid documentation for those who are unfamiliar mermaid: https://mermaid.js.org/intro/n00b-gettingStarted.html
> * and I would include a simple sample flowchart that people can start to tweak (or delete), instead of greeting them with an empty notepad and an empty preview.

Thanks for the feedback @zsviczian! makes sense, I will add these

**Comment by ad1992** [comment]
> 🎉
> 
> Haven't checked code yet, but some notes on the UI:
> 
> image

Have made most of the tweaks suggested apart from the 👇🏻 
1. I think we shouldn't be doing `full-width` to keep it consistent with other dialogs e.g. `SaveTo`, and `Help` dialogs ?
2. Border I have used a slightly lighter version of what was in the design as the design one was looking dark, Let me know if it is fine

**Comment by zsviczian** [comment]
If I insert a non-flowchart mermaid diagram it is displayed as an SVG image. This is great! However, if I want to go back to modify the mermaid script for that SVG, that does not seem possible. I can click the mermaid tool again, however if in the mean time I insert a second diagram, currently there is no going back to the first one.

1) Would be great to offer an option to save the mermaid script with the image (maybe make it into and ExcalidrawMarmaidElement that could be inherited from ExcalidrawImageElement with the addition that it has a text or description property that holds the mermaid script).

2) If 1. is implemented, it would be very helpful to allow inserting flowcharts as element.type="mermaid" as well. The benefit of this would be ability to add boilerplate flowcharts to the stencil library.

3) One more idea is to save mermaid scripts within a diagram as a history item or maybe in AppState and offer a backwards/forward navigation between scripts in the dialog window, saving these scripts to browser cache.

**Comment by dwelle** [comment]
> 1. Would be great to offer an option to save the mermaid script with the image (maybe make it into and ExcalidrawMarmaidElement that could be inherited from ExcalidrawImageElement with the addition that it has a text or description property that holds the mermaid script).

Yes, we were considering something along those lines.

> If 1. is implemented, it would be very helpful to allow inserting flowcharts as element.type="mermaid" as well. The benefit of this would be ability to add boilerplate flowcharts to the stencil library.

That was the concern with (1), as we implement more and more chart types to be converted directly to excalidraw, we'd face this problem.

Ideally we come up with a way that keeps both the mapping to original mermaid source, and keeps the output as editable elements. There are some ideas on that front, but there'd still likely be a disconnect between what's on canvas and the mermaid source because in the meantime the user could have edited the elements — unless we support two-way transformation (excalidraw back to mermaid).

**Comment by DanielJGeiger** [comment]
> > 1. Would be great to offer an option to save the mermaid script with the image (maybe make it into and ExcalidrawMarmaidElement that could be inherited from ExcalidrawImageElement with the addition that it has a text or description property that holds the mermaid script).
> 
> Yes, we were considering something along those lines.

@zsviczian Good idea to keep the mermaid source with the diagram.  In my experience doing so with MathJax (#6037), text in an svg may not render the same as text rendered directly to the canvas.  (Especially at zoom levels other than 100%.)

> 
> > If 1. is implemented, it would be very helpful to allow inserting flowcharts as element.type="mermaid" as well. The benefit of this would be ability to add boilerplate flowcharts to the stencil library.
> 
> That was the concern with (1), as we implement more and more chart types to be converted directly to excalidraw, we'd face this problem.

@dwelle You raise a good point.  Your idea sounds like a good use-case for the subtypes system I developed in #6037.  Integrating @zsviczian's ideas, I would suggest an `ExcalidrawChartElement` type inheriting from `ExcalidrawTextElement` with `element.subtype = "mermaid"` (and `element.subtype = "otherchart"` etc.)  If chart elements inherited from text elements instead of image elements, that would avoid the text rendering quirks I once experienced in #6037 (it was #2993 at the time).

**Comment by ad1992** [comment]
> @ad1992
> 
> summing up fixes we need to do before WIP release:
> 
> * [x]  fix safari
> * [x]  fix binding gaps
> * [x]  paste in viewport center
>   Right now it pastes relative to `[0,0]`, I assume because that's what get from the mermaid parser. So related to previous issue because we'll either need to change the positioning in the mermaid package, or renormalize the positions post parse.
> * [x]  debounce the preview
>   When the diagram is more complex or on slow CPUs the parsing + renderings takes too long resulting in lag when typing. Let's debounce ~100-200ms.

All the issues should be fixed now.
Instead of debouncing I am using `useDeferredValue` hook as this is a good use case for this.

File: src/css/styles.scss
Comment (dwelle) [inline src/css/styles.scss:617]: this will make it not scrollable on small mobile screens

image

**Comment by ad1992** [comment]
> It seems the mermaid chunks are super granular, and also being loaded right on init load, not when the dialog is being opened.
> 
> Also we're not splitting our vendor or other chunks too greedily, resulting in too many requests (83 vs 45 in production)
> 
> image

Btw I assume you were talking about package here and not excalidraw app as that's taken care by Vite and I don't see any difference between production and preview

## Stats

**Production**

Requests - 31
Time - 2.73 sec

**Preview**
Requests - 30
Time - 2.38 sec

And all mermaid stuff  is loaded lazily as well once you open the dialog

Screenshot 2023-09-05 at 1 37 17 PM

So changing the bundling strategy will be needed for the package and not the app as that's taken care by Vite itself so not a blocker for the app but the package

File: src/css/styles.scss
Comment (ad1992) [reply src/css/styles.scss:617]: Exactly why do we want it to be scrollable? Hence I have changed the CSS a bit as well so it works on most of the mobile screens. 
We shouldn't anyways have horizontal scroll for the below reasons
1. Its a bad UX and not intuitive at all
2. We can just move one of the tools inside "more tools" or on the side to support it in more small small screens up to 320px which is given by A11y guidelines.

Right now with this change I am hoping it would work well in most of the screens but as suggested I would like to move 1-2 tools inside more so UX is less cluttered and that we can do for smaller screens only after certain threshold and not for mobile layout

Comment (dwelle) [reply src/css/styles.scss:617]: Before we do a mobile redesign the menu at least has to be _usable_ on small screens. Bad UX is definitely better than "not possible to select a tool at all".

Comment (ad1992) [reply src/css/styles.scss:617]: Well, what you are talking about is just ~45px (below 365 to 320px) what we need to support, which is very small range and I am not sure if it's a standard mobile screen, but in any case what issue do you see If we move one of the tools inside more tools or on the side for this 45px range?

For other sizes, the current UX is much better and intuitive

**Comment by ad1992** [comment]
> > It seems the mermaid chunks are super granular, and also being loaded right on init load, not when the dialog is being opened.
> > Also we're not splitting our vendor or other chunks too greedily, resulting in too many requests (83 vs 45 in production)
> > image
> 
> Btw I assume you were talking about package here and not excalidraw app as that's taken care by Vite and I don't see any difference between production and preview
> 
> ## Stats
> **Production**
> 
> Requests - 31 Time - 2.73 sec
> 
> **Preview** Requests - 30 Time - 2.38 sec
> 
> And all mermaid stuff is loaded lazily as well once you open the dialog
> 
> Screenshot 2023-09-05 at 1 37 17 PM
> So changing the bundling strategy will be needed for the package and not the app as that's taken care by Vite itself so not a blocker for the app but the package

Also to edit, the package requests are fine too so not related to package as well. Maybe you meant "Local", but local is the same as well in master so not sure which two builds you were exactly comparing ?

**Comment by ad1992** [comment]
> > > It seems the mermaid chunks are super granular, and also being loaded right on init load, not when the dialog is being opened.
> > > Also we're not splitting our vendor or other chunks too greedily, resulting in too many requests (83 vs 45 in production)
> > > image
> > 
> > 
> > Btw I assume you were talking about package here and not excalidraw app as that's taken care by Vite and I don't see any difference between production and preview
> > ## Stats
> > **Production**
> > Requests - 31 Time - 2.73 sec
> > **Preview** Requests - 30 Time - 2.38 sec
> > And all mermaid stuff is loaded lazily as well once you open the dialog
> > Screenshot 2023-09-05 at 1 37 17 PM
> > So changing the bundling strategy will be needed for the package and not the app as that's taken care by Vite itself so not a blocker for the app but the package
> 
> Also to edit, the package requests are fine too so not related to package as well. Maybe you meant "Local", but local is the same as well in master so not sure which two builds you were exactly comparing ? I have also updated the prev comment with stats of local and package

To clarify further it was only the first request where the number has increased to ~84 and it is for SW so it can cache for offline usage. All future requests are similar to prod and mermaid is lazily loaded as well.
The number of requests is huge for the first time due to Mermaid integration but it's non-blocking as it runs in the background so won't impact user experience. The bandwidth and therefore the cost would get affected due to the first request when we release it.

**Comment by ad1992** [comment]
> I like the feature but i have no idea about mermaid syntax so any resources to learn?

@lash0000 you can check the [`flowcharts`](https://mermaid.js.org/syntax/flowchart.html) as mentioned in the mermaid popup to get started

File: src/css/styles.scss
Comment (ad1992) [reply src/css/styles.scss:617]: @dwelle in case you missed the above message ^

Comment (dwelle) [reply src/css/styles.scss:617]: let's do this and hope that there aren't many screens smaller than that.

> in any case what issue do you see If we move one of the tools inside more tools or on the side for this 45px range

We can , but I'd do it conditionally, only if the screen size won't fit the tools. And I'm not sure which tool we'd hide there (diamond?).

**Comment by dwelle** [comment]
There's a regression in styling the right-side tool menu on mobile:

prod:
image
this pr:
image

File: src/css/styles.scss
Comment (ad1992) [reply src/css/styles.scss:617]: I think since one of the tools will go inside the dropdown the eraser should go as that's the tool on extreme right when the width is exceeding so might be more intuitive to push the eraser instead of picking one of the shapes tools.
wdyt?

Comment (ad1992) [reply src/css/styles.scss:617]: For now, we would probably keep as is as this consideration is just for ~45px as mentioned here - https://github.com/excalidraw/excalidraw/pull/6920#discussion_r1315564739 and I am not sure if this is something we need to be worried about so we can ship it as is and later push one of the items to dropdown if this becomes a concern

Comment (dwelle) [reply src/css/styles.scss:617]: yeah, in the end we decided to leave as is as it affects only a very small mobile screen, and long-term we plan to redesign mobile completely.

Comment (ad1992) [reply src/css/styles.scss:610]: Alright as I was fixing the regression of mobile toolbar, I looked into this again and I have made it work for smaller screens up to 320px 


Screenshot 2023-09-21 at 5 52 01 PM

File: src/element/dragElements.ts
Comment (dwelle) [inline src/element/dragElements.ts:71]: this whole file should be a separate PR

File: src/types.ts
Comment (dwelle) [inline src/types.ts:96]: the mermaid popup should use `appState.openDialog` as it's not a tool (it'll also make it not persist across reloads which is the correct behavior). Was there any reason why use made it into a tool (apart from the fact it's in the toolbar)?

**Comment by zsviczian** [comment]
I don't mind contributing directly to your fork, @ad1992.

Regarding storing the diagram definition text in the customData of the generated image, it may not be the optimal approach, but it appeared to be the simplest way to implement the feature I wanted. Now, when I select a Mermaid image of a diagram (assuming it's an SVG and not Excalidraw elements), the editor will always show the original script. This enables me to easily make modifications and regenerate the diagram, and I believe it will greatly enhance the usability of this feature.

**Comment by ad1992** [comment]
> I hope you don't mind me contributing directly to your fork, @ad1992.
> 
> Regarding storing the diagram definition in `customData` of the generated image, it may not be the best approach, but it appeared to be the simplest way to implement the feature I wanted. Also, I was thinking, that eventually all the diagram types are going to be supported by mermaid-to-excalidraw, thus this is not a permanent feature and would therefore likely not warrant a new property on ExcalidrawImageElement.
> 
> With this UX improvement, when the user selects a Mermaid generated SVG image on the canvas, the editor will display the diagram definition matching the selected diagram. This enables the user to easily make modifications and regenerate the diagram. I believe this will greatly enhance the usability of the mermaid feature when working with embedded mermaid diagrams.

Hi @zsviczian, was on vacation so couldn't get back on this. Thanks for the enhancement!

However, I would really appreciate if we discuss the approach/scope first and based on that we push to the current PR or raise a new PR against the current PR so we all are on the same page and the scope of the current PR doesn't get impacted. I would prefer raising a separate PR unless its related to fixes for the main PR so it is easy to review and test. Also tests need to be added as well for the changes. Once it looks good we can merge it to the main PR or push as an enhancement later too.

This is a good enhancement, but there are some bugs eg If I select an existing mermaid image -> open the mermaid editor -> edit the mermaid syntax-> close the editor. Now if I again select the same old image, it would still show the updated data which isn't right. So will need to test and also add the tests for it as well. 

Also, I am not sure if `customData` is the right attribute for this as the host could pass their own custom data as well and we need to make sure that's not impacted so maybe we add a new property, will need to discuss that. 
Hence I think it would be better if you could pull the commits and raise a separate PR against this one and we can finalize the approach, test it out, and add tests for this in the PR, we can merge to this PR or add as an enhancement right after its shipped.

**Comment by zsviczian** [comment]
sure - happy to handle in a separate PR. 
As for customData. If host app uses customData the responsibility is on them to update customData such that they preserve existing data contents (if they so prefer). I wouldn't be concerned about that. Because eventually we'd want all mermaid diagram types processed into Excalidraw shapes I think a temporary solution like customData is better, but I leave it up to you. The key for me from a UX perspective is that when I click on a Mermaid SVG I'd prefer being able to edit it.

**Comment by ad1992** [comment]
> sure - happy to handle in a separate PR. As for customData. If host app uses customData the responsibility is on them to update customData such that they preserve existing data contents (if they so prefer). I wouldn't be concerned about that. Because eventually we'd want all mermaid diagram types processed into Excalidraw shapes I think a temporary solution like customData is better, but I leave it up to you. The key for me from a UX perspective is that when I click on a Mermaid SVG I'd prefer being able to edit it.

May be but currently we aren't communicating that `custom data` can be manipulated by us and host needs to handle it so let's think through it before finalizing. The custom data was added with the intention so that host can add whatever custom data they want to support. Lets discuss this once you open the PR.

**Comment by ad1992** [comment]
@zsviczian as discussed since we agreed on moving the changes to separate PR you can find your commits in my [backup branch](https://github.com/excalidraw/excalidraw/tree/aakansha/mermaid-dupl) incase you don't have those commits. I will be removing those commits from this branch.

**Comment by zsviczian** [comment]
sure. thanks.

File: src/element/dragElements.ts
Comment (ad1992) [reply src/element/dragElements.ts:69]: Yes it's already in https://github.com/excalidraw/excalidraw/pull/6891, got pushed in this PR mistakenly

File: src/types.ts
Comment (ad1992) [reply src/types.ts]: Yes the only reason to put it under tools was its in the toolbar

File: excalidraw-app/debug.ts
Comment (dwelle) [inline excalidraw-app/debug.ts:134]: nice that someone in the DT repo for the `debug` package started polluting the global namespace or whatever is happening here :(

File: src/components/Actions.tsx
Comment (dwelle) [inline src/components/Actions.tsx:231]: we don't wanna highlight the more toools button or the mermaid button since it's a dialog and highlighting would serve no purpose as it'd be highlighted only while the dialog is open.

File: src/components/App.tsx
Comment (dwelle) [inline src/components/App.tsx:7797]: replace with `selectedElementIds: makeNextSelectedElementIds(selectedElementIds, prevState)` to keep the object stable when appliacble

Comment (dwelle) [reply src/components/App.tsx:7797]: actually also for `selectedGroupIds` (use the same function, we'll need to rename later)

File: src/components/App.tsx
Comment (dwelle) [inline src/components/App.tsx:7783]: checking for `!isBoundToContainer(ele)` should be done here as well. In other words, the `selectedElementIds[ele.id] = true` should be assigned only once at the top, irrespective of groups

File: src/components/App.tsx
Comment (dwelle) [inline src/components/App.tsx:7786]: as far as I know this is not correct as we don't select all groups if you're drilling into a parent group(s).

---

The whole selection thing is tricky, and even all this is not enough as there are frames and such. So either we need to rewrite this completely, or I'd just inline it for the mermaid's sake and leave this function for later.

File: src/data/__snapshots__/transform.test.ts.snap
Comment (dwelle) [inline src/data/__snapshots__/transform.test.ts.snap:1338]: looks sus

**Comment by ad1992** [comment]
@excalibot trigger release

File: src/components/MermaidToExcalidraw.tsx
Comment (dwelle) [reply src/components/MermaidToExcalidraw.tsx:118]: FWIW let's remove the `loading` flag completely and add a state for the `mermaidToExcalidrawLib` lib instead, and use that as the loading state

File: src/components/icons.tsx
Comment (dwelle) [inline src/components/icons.tsx:1659]:
```suggestion
    fill="strokeColor"
```

File: src/components/icons.tsx
Comment (dwelle) [inline src/components/icons.tsx:1671]: let's drop these and use `modifiedTablerIconProps` as options object, and put the `strokeWidth` onto a parent `<g>` elem.

File: src/components/MermaidToExcalidraw.tsx
Comment (dwelle) [inline src/components/MermaidToExcalidraw.tsx:158]:
```suggestion
        parent.style.background = "var(--default-bg-color)";
```

File: src/components/MermaidToExcalidraw.tsx
Comment (dwelle) [inline src/components/MermaidToExcalidraw.tsx:63]: move styling to the scss file

File: src/components/MermaidToExcalidraw.tsx
Comment (dwelle) [inline src/components/MermaidToExcalidraw.tsx:226]:
```suggestion
              border: "1px solid var(--color-border-outline)",
```

also move the whole styling to the scss file

File: src/components/MermaidToExcalidraw.scss
Comment (dwelle) [inline src/components/MermaidToExcalidraw.scss:49]:
```suggestion
      border: 1px solid var(--color-border-outline);
```

File: src/components/MermaidToExcalidraw.scss
Comment (dwelle) [inline src/components/MermaidToExcalidraw.scss:48]:
```suggestion
      border-radius: var(--border-radius-lg);
```

File: src/components/MermaidToExcalidraw.scss
Comment (dwelle) [inline src/components/MermaidToExcalidraw.scss:28]:
```suggestion
        border-radius: var(--border-radius-lg);
```

File: src/components/MermaidToExcalidraw.tsx
Comment (dwelle) [inline src/components/MermaidToExcalidraw.tsx:161]: logging the error would unnecessarily pollute our sentry tracker

File: src/components/MermaidToExcalidraw.tsx
Comment (dwelle) [inline src/components/MermaidToExcalidraw.tsx:229]:
```suggestion
            onChange={(event) => setText(event.target.value.trim())}
```

File: src/components/MermaidToExcalidraw.tsx
Comment (dwelle) [inline src/components/MermaidToExcalidraw.tsx:125]:
```suggestion
      }

      if (!deferredText) {
        resetPreview();
        return;
      }

```

File: src/components/App.tsx
Comment (ad1992) [reply src/components/App.tsx:7795]: Actually I just realized we no more need this API at all since we are using `addElementsFromPasteOrLibrary` hence I am removing this API altogether

File: src/data/__snapshots__/transform.test.ts.snap
Comment (ad1992) [reply src/data/__snapshots__/transform.test.ts.snap:1608]: This is fine, we shouldn't be updating arrow height when bound text height exceeds hence this snapshot was updated (check textElement.ts diff)

File: src/components/MermaidToExcalidraw.tsx
Comment (ad1992) [reply src/components/MermaidToExcalidraw.tsx:118]: true `loading` shouldn't be used for updating text area. 
I have moved the `loading` state inside `renderExcalidrawPreview` instead. However I am not sure what you mean by removing the flag and adding state for the lib ? Do you mean just renaming loading State to `isMermaidLibraryLoaded` ?

File: src/components/MermaidToExcalidraw.tsx
Comment (ad1992) [reply src/components/MermaidToExcalidraw.tsx:206]: This will prevent the user from adding whitespace which is not the intention. Trimming can be done instead on deferred text

File: src/components/MermaidToExcalidraw.tsx
Comment (dwelle) [reply src/components/MermaidToExcalidraw.tsx]: it's per spec and we use it (or should) everywhere else. But right now it's kind of a mess and there's not much consistency across the board, so let's use `--dialog-border-color` instead (to fix dark mode).

File: src/components/MermaidToExcalidraw.scss
Comment (dwelle) [reply src/components/MermaidToExcalidraw.scss:49]: ditto, `--dialog-border-color`

**Review by dwelle — APPROVED** [review]

**Comment by dwelle** [comment]
We're good to go! 🔥 🚀

**Pull request merged**
